### ISSUE TITLE
Support for Ruby 3.1

### ISSUE THREAD

**Issue by januszm** [aws/elastic-beanstalk-roadmap#232]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave "+1" or "me too" comments. They generate extra noise for issue followers and do not help prioritize the request.

**Tell us about your request**
Since the first ‘teeny’ release of Ruby 3.1  ->  3.1.1 is out, it should be safe to use it in production. Please add new Ruby 3.1 solution stack to Elastic Beanstalk platform.

**Is this request specific to an Elastic Beanstalk [platform]
Ruby

**Attachments**
https://www.ruby-lang.org/en/news/2022/02/18/ruby-3-1-1-released/

**Comment by januszm** [comment]
@1tonyca  @zzzhu0192  @baiyli Any updates on this? Ruby 3.1.0 Was released back in December 2021, 3.1.1 (first "teeny" version, definitely safe for production use) was released in February 2022. 

How much time does AWS EB teams need to introduce support for the new language version? After all, there is no revolution in the syntax or problematic breaking changes, which would typically cause such delays (PS. I represent a paying customer who has many applications on the EB platform)

**Comment by seuros** [comment]
> How much time does AWS EB teams need to introduce support for the new language version?

EB team seem focused on EOL technology only.   
By the time they will give us Ruby 3.1 as default, Ruby 4 will be around.

You can still install ruby 3.1 manually using rbenv build.

**Comment by januszm** [comment]
> You can still install ruby 3.1 manually using rbenv build.

@seuros Could you please share how you managed to install and use newer Ruby in EB? As  far as I remember, the server lacks rbenv/ruby-build, which is necessary for the installation. Unfortunately, I don't remember if it was possible to add ruby-build on a running server or not last time I tried.

UPDATE: nvm, I was able to install ruby-build with:
```sh
sudo su -
git clone https://github.com/rbenv/ruby-build.git
PREFIX=/opt/elasticbeanstalk/.rbenv ./ruby-build/install.sh
```
However it's quite impractical, I can't imagine running this on each immutable deployment. Another option is to use custom AMI but it kind of defeats the purpose of using a managed platform like this.

Perhaps open-sourcing the EB Ruby Platform generation code and letting the community add updates would solve this issue?

**Comment by kokogakayui** [comment]
Hi ! Anything new about this ?

**Comment by ashkulz** [comment]
The Ruby 3.2 release is just around the corner.

**Comment by anatoljevich** [comment]
Come on, guys. EB still supports only ruby-3.0.5 while Codebuild ruby-2.6, 2.7 and 3.1.
How do you think should I deploy rails applications?

**Comment by ashkulz** [comment]
I talked to AWS support, and they recommended to use Docker. This is on their "features roadmap" without any ETA :man_shrugging: I also reached out our account rep about this but he too hasn't responded.

**Comment by januszm** [comment]
Currently this platform is left without support and I'm thinking whether it makes sense to create a more "community driven" future Ruby platform using instructions at: https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/custom-platforms.html#custom-platforms-sample
after all they use Packer from HashiCorp to build platforms

Meanwhile, this can be used to install Ruby 3.2

```
git clone https://github.com/rbenv/ruby-build.git "$(rbenv root)"/plugins/ruby-build
yum -y install libyaml-devel libffi-devel readline-devel
rbenv install 3.2.1
...
Installing ruby-3.2.1...
Installed ruby-3.2.1 to /opt/elasticbeanstalk/.rbenv/versions/3.2.1
```

But I noticed that the server goes down some minutes after new Ruby is installed.

**Comment by januszm** [comment]
@ashkulz 
> I talked to AWS support, and they recommended to use Docker. 

I also talked to AWS support and they said the same thing,  but this is not an option for me, at least not  a viable option. 

Switching to Docker opens a whole new category of problems (e.g. observability, how are you going to run a DataDog agent with your app, not to mention the enormous amount of time simply wasted on dockerizing the app for AWS and maintaining it later).
Development teams will now need to waste time on this instead of focusing on building business logic for their apps.

**Comment by kokogakayui** [comment]
I was really relunctant to switch to Docker, but after months waiting for Ruby 3.1 platform, and the release of Ruby 3.2, we finally took the decision and migrated our app. To be honest, it was a pretty straightforward and easy process, and we did gain a lot in terms of flexibility and freedom. No need to wait for AWS to update the platform dependencies anymore, we can now do it when and if we want. And it is really not that hard to maintain. 
I feel like AWS is allocating less and less energy into those opinionated platforms, so no regrets here, we would probably have to do it at some point anyways.
